x-pack/plugin/otel-data/src/main/resources/component-templates/traces-otel@mappings.yaml

```diff
@@ -0,0 +1,67 @@
+---
+version: ${xpack.oteldata.template.version}
+_meta:
+  description: Default mappings for OpenTelemetry traces
+  managed: true
+template:
+  mappings:
+    date_detection: false
+    properties:
+      trace_id:
+        type: keyword
+      trace.id:
+        type: alias
+        path: trace_id
+      span_id:
+        type: keyword
+      span.id:
+        type: alias
+        path: span_id
+      trace_state:
+        type: keyword
+      parent_span_id:
+        type: keyword
+      parent.id:
+        type: alias
+        path: parent_span_id
+      trace_flags:
+        type: byte
```

Comment:
type `byte` is probably not enough. See https://github.com/open-telemetry/opentelemetry-collector/blob/main/pdata/internal/data/protogen/trace/v1/trace.pb.go#L400-L421

Reply:
What type do we want to use here?

The collector uses unsigned int, but we don't have unsigned int in Elasticsearch.

I see 2 options:
- integer
  - Pro: less space - we only use 32bit
  - Con: Not future proof. The comment on span indicates that the 10 least significant bits are used, so that means, we can store those values in a signed integer. But then it also says: `Readers MUST NOT assume that bits 10-31 (22 most significant bits) will be zero.` - now that means this may already overflow if we store it in a signed int. Furthermore, Logs has the same field (I'll update that field as well), and [that comment specifically says](https://github.com/open-telemetry/opentelemetry-collector/blob/main/pdata/internal/data/protogen/logs/v1/logs.pb.go#L406-L411): `24 most significant bits are reserved`.

- long
  - Con: it's a 64bit field, so we waste 32 bit
  - Pro: if most significant bits are getting used in the collector uint, we can still store those values. So this addresses the `cons` with integer.

I'd note that an alternative approach could be to parse these flags in the collector (probably in the es exporter) and store those in separate fields. Currently there are 2 things there:
- The 8bit trace-flag from the W3C traceparent header - [that has currently 2 bits defined](https://www.w3.org/TR/trace-context-2/#sampled-flag) - sampled and  random trace id.
- In case of a span, whether the parent is remote, which we just removed: https://github.com/elastic/opentelemetry-dev/pull/364.

I'm not sure parsing is the way to go, the reason I bring it up is because these bitflag in its current form may not be the most useful. I don't really see what we'd do with it as is.

Overall, I'd suggest keeping the field and using type `long`.

What do others think? @felixbarny.

Reply:
I did a quick test to see what happens if those most significant bits are set and we use integer.

Elasticsearch won't let it overflow and set it to a negative number - so even if in theory this would work, Elasticsearch prevents us doing it - which kind of makes sense. So this really makes me think we should go with long here.

```
{
  ".ds-foobar-2024.08.08-000001": {
    "mappings": {
      "_data_stream_timestamp": {
        "enabled": true
      },
      "properties": {
        "@timestamp": {
          "type": "date"
        },
        "intfield": {
          "type": "integer"
        }
      }
    }
  }
}
```

```
POST foobar/_doc
{
  "@timestamp":  "2024-08-05T10:43:36.250Z",
  "intfield": 4294967295
}
```


Response:

```
{
  "error": {
    "root_cause": [
      {
        "type": "document_parsing_exception",
        "reason": "[1:59] failed to parse field [intfield] of type [integer] in document with id 'ZDKrMpEBZ0KOL9if4h2U'. Preview of field's value: '4294967295'"
      }
    ],
    "type": "document_parsing_exception",
    "reason": "[1:59] failed to parse field [intfield] of type [integer] in document with id 'ZDKrMpEBZ0KOL9if4h2U'. Preview of field's value: '4294967295'",
    "caused_by": {
      "type": "x_content_parse_exception",
      "reason": """[1:69] Numeric value (4294967295) out of range of int (-2147483648 - 2147483647)
 at [Source: (byte[])"{  "@timestamp": "2024-08-05T10:43:36.250Z",  "intfield": 4294967295}
"; line: 1, column: 69]""",
      "caused_by": {
        "type": "input_coercion_exception",
        "reason": """Numeric value (4294967295) out of range of int (-2147483648 - 2147483647)
 at [Source: (byte[])"{  "@timestamp": "2024-08-05T10:43:36.250Z",  "intfield": 4294967295}
"; line: 1, column: 69]"""
      }
    }
  },
  "status": 400
}
```

Reply:
`long` might be the only option if we want to be compliant and avoid fixing this down the line when otel decides to start using the significant bits.

Reply:
Trace flags is an 8 bit field: https://www.w3.org/TR/trace-context/#trace-flags
Why do we need another data type than byte?

Reply:
> Trace flags is an 8 bit field: https://www.w3.org/TR/trace-context/#trace-flags

That is just one part of what's stored in the [`flags` field linked above](https://github.com/open-telemetry/opentelemetry-collector/blob/main/pdata/internal/data/protogen/trace/v1/trace.pb.go#L400-L421). The first 8 bits of the `flags` on the span is the 8bit trace flags from W3C, then the next 2 bits are the `is remote parent` thing (with it's 3 values), and the remaining 22 bits are not used, but as the comment indicates may not be zero.

So this is a flag, which stores another flag in its first part :)

Reply:
> `long` might be the only option if we want to be compliant and avoid fixing this down the line when otel decides to start using the significant bits.

Not only when they start using the significant bits. Sure, that's a clear problem then. But even if those are not zeroed, as the comment indicates (even if there is no definition for those yet), then Elasticsearch will already reject those numbers.

Reply:
Oh, I see. Thanks for the clarification. While ES does have an [`unsigned_long`](https://www.elastic.co/guide/en/elasticsearch/reference/current/number.html) data type, there's no `unsigned_integer`. Therefore, `long` seems to be the best fit.

Reply:
Thanks for the input - seems we have an agreement.

Changed it to `long`: https://github.com/elastic/elasticsearch/pull/111091/commits/a6baf4644924ea4929ae2ab58ebac0b78f936e62

And a PR to document this in our spec: https://github.com/elastic/opentelemetry-dev/pull/368

Reply:
Update: we decided to drop this field for now, so it's removed from the mappings. See: https://github.com/elastic/opentelemetry-dev/pull/371